ME-48 Создание базового мультиплеера

BeetleRat opened the pull request:
Dev:

- Отправка сообщения о создании хоста на файловый сервер с указанием ip и порта
- Подключение к хосту по ip и порту
- Спавн клиента в сцене хоста
- Отключение контроллеров других пользователей
- Установка клиентского аватара
- Добавление отображаемого имени пользователя над аватаром игрока

Problem:

- `NetworkManager#offlineScene` больше не используется. Мы в ручную настраиваем переключение между сценами. Имя сцены, в которую должен попадать игрок при отключении от комнаты, храниться в `NetworkManager#onlineScene`. Однако интерактивность кнопки возвращения в офлайн сцену определяется через `NetworkManager#offlineScene`
- `yield return new WaitUntil(() => AvatarDownloading().IsCompleted);` вызывает метод `AvatarDownloading` несколько раз
- Скрипт `UILoadingForm` отключает объект в котором находится и больше не может вызываться

Fix:

- Определять интерактивность кнопки возвращения в офлайн сцену через `NetworkManager#onlineScene`
- Заранее объявить Task вызова метода и дожидаться его окончания
- Для префаба `LoadingForm` cоздать подобъект содержащий отображаемую часть UI. Скрывать и отображать подобъект. А объект на котором присутствует скрипт всегда оставлять активным

BeetleRat committed:
ME-48 Исправление кнопки возвращения на офлайн сцену

Problem:

- `NetworkManager#offlineScene` больше не используется. Мы в ручную настраиваем переключение между сценами. Имя сцены, в которую должен попадать игрок при отключении от комнаты, храниться в `NetworkManager#onlineScene`. Однако интерактивность кнопки возвращения в офлайн сцену определяется через `NetworkManager#offlineScene`

Fix:

- Определять интерактивность кнопки возвращения в офлайн сцену через `NetworkManager#onlineScene`

BeetleRat committed:
ME-48 Создание базового мультиплеера

Dev:

- Отправка сообщения о создании хоста на файловый сервер с указанием ip и порта
- Подключение к хосту по ip и порту

BeetleRat committed:
ME-48 Создание базового мультиплеера

Dev:

- Спавн клиента в сцене хоста
- Отключение контроллеров других пользователей

BeetleRat committed:
ME-48 Создание базового мультиплеера

Dev:

- Установка клиентского аватара

BeetleRat committed:
ME-48 Исправление повторной загрузки аватаров

Problem:

- `yield return new WaitUntil(() => AvatarDownloading().IsCompleted);` вызывает метод `AvatarDownloading` несколько раз

Fix:

- Заранее объявить Task вызова метода и дожидаться его окончания

BeetleRat committed:
ME-48 Создание базового мультиплеера

Dev:

- Добавление отображаемого имени пользователя над аватаром игрока

BeetleRat committed:
ME-48 Исправление префаба LoadingForm

Problem:

- Скрипт отключает объект в котором находится и больше не может вызываться

Fix:

- Создать подобъект содержащий отображаемую часть UI. Скрывать и отображать подобъект. А объект на котором присутствует скрипт всегда оставлять активным

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/MVNetworkManager.cs:
```diff
@@ -129,8 +166,21 @@ public override void Awake()
         public override void OnStartHost()
         {
             base.OnStartHost();
+            AppLogger.Log("After start host");
```
Давай все же уберем эти логгеры.

BeetleRat replied:
Fixed.

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/Player/AvatarPlayer/NetworkAvatarPlayer.cs:
```diff
@@ -0,0 +1,74 @@
+            avatarPrefabInfo.ForGender = avatar.AvatarGender;
+            playerAvatar.AvatarComponent.CreatePlayerFromAvatar(avatarPrefabInfo);
+
+            NetworkTransformReliable ntr = GetComponent<NetworkTransformReliable>();
```
ntr - плохой нейминг

BeetleRat replied:
Fixed.

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/Player/Spawn/NetworkPlayerInstantiation.cs:
```diff
@@ -0,0 +1,219 @@
+
+        private void OnHostStarted()
+        {
+            AppLogger.Log("Host Started");
```
Убери логгеры

BeetleRat replied:
Fixed

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/Player/Spawn/NetworkPlayerInstantiation.cs:
```diff
@@ -0,0 +1,219 @@
+            SendRegistrationMessageToServer();
+        }
+        
+        /// Клиент не имеет прямого доступа к серверным методам и объектам.
```
Лишний комментарий. Он есть в `ClientRegistrationMessage`. Если так хочется, унеси его внутрь метода, а не делай документацию.

BeetleRat replied:
Fixed.

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/Types/Client/ClientSpawnData.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace NetworkCore.MirrorNetworking.Types.Client
```
Перенеси в `NetworkCore.MirrorNetworking.Containers`

BeetleRat replied:
Fixed.

metaverseguap commented on Assets/ApplicationContent/NetworkCore/MirrorNetworking/Utils/NetworkManagerExtensions.cs:
```diff
@@ -134,18 +162,44 @@ public static void BecomeHost(this MVNetworkManager connection, SceneInfo scene)
+        /// <para>Данный метод используется для заранее созданных префабов</para>
+        /// </summary>
+        /// <param name="connection"><see cref="MVNetworkManager"/></param>
+        /// <param name="prefab">спавнемый префаб</param>
```
Должен иметь NetwokIdentity в Unity Editor

BeetleRat replied:
Fixed.

metaverseguap commented on Assets/ApplicationContent/NetworkCore/ServerInteraction/API/HostsAPI.cs:
```diff
@@ -64,24 +66,22 @@ public IList<HostInfo> GetHostsBySceneName(string sceneName)
         /// <param name="sceneName">имя сцены, хостом которой стал пользователь</param>
-        /// <returns>true, если запрос завершился успешно</returns>
-        public bool BecomeAHost(string url, string sceneName)
+        /// <returns>адрес созданного хоста</returns>
```
Или null если не удалось создать

BeetleRat replied:
Fixed

metaverseguap commented on Assets/ApplicationContent/NetworkCore/Utils/IPUtils.cs:
```diff
@@ -61,5 +82,53 @@ public static bool IsURLValid(string url)
+        /// </summary>
+        /// <param name="range">количество портов</param>
+        /// <returns>свободный порт или null, если все порты заняты</returns>
+        public static int? GetAvailableUDPPort(int range)
```
GetAvailablePortUDP

BeetleRat replied:
Fixed

metaverseguap commented on Assets/ApplicationContent/NetworkCore/Utils/IPUtils.cs:
```diff
@@ -61,5 +82,53 @@ public static bool IsURLValid(string url)
+                IPGlobalProperties.GetIPGlobalProperties()
+                    .GetActiveUdpListeners()
+                    .Select(p => p.Port)
+                    .OrderBy(p => p) // В debug проще смотреть на отсортированные значения 
```
Для релизной версии лучше закомментировать

BeetleRat replied:
Fixed.

BeetleRat committed:
ME-48 Исправление замечаний MR

Fix:

- Убраны Logger-ы из `MVNetworkManager` и `NetworkPlayerInstantiation`
- Изменена иерархия в namespace `NetworkCore.MirrorNetworking.Containers`
- Добавлена проверка на null для результата `HostsAPI#BecomeAHost`
- В `IPUtils#GetAvailablePortUDP` закомментирована сортировка портов
- Добавлены комментарии
- Произведен рефакторинг
